XenAdmin/Commands/ShutDownHostCommand.cs

```diff
@@ -143,21 +145,40 @@ protected override string ConfirmationDialogText
 
                     if (h.Connection.ResolveAll(h.resident_VMs).Exists(v => v.HciWarnBeforeShutdown()))
                         hciHosts.Add(h);
+
+                    if (Helpers.HostIsMaster(h) && h.Connection.Cache.HostCount > 1)
+                        poolMasters.Add(h);
                 }
 
+                StringBuilder sb = new StringBuilder();
+
                 if (hciHosts.Count > 0)
-                    return hciHosts.Count == 1
-                        ? string.Format(Messages.CONFIRM_SHUTDOWN_HCI_WARN_SERVER, hciHosts[0].Name())
-                        : string.Format(Messages.CONFIRM_SHUTDOWN_HCI_WARN_SERVERS, string.Join("\n", hciHosts.Select(h => h.Name())));
-
-                if (hasRunningVMs)
-                    return hosts.Count == 1
-                        ? string.Format(Messages.CONFIRM_SHUTDOWN_SERVER, hosts[0].Name())
-                        : Messages.CONFIRM_SHUTDOWN_SERVERS;
-
-                return hosts.Count == 1
-                    ? string.Format(Messages.CONFIRM_SHUTDOWN_SERVER_NO_VMS, hosts[0].Name())
-                    : Messages.CONFIRM_SHUTDOWN_SERVERS_NO_VMS;
+                    sb = hciHosts.Count == 1

```

Comment:
The Conditional/Ternary operator returns one of two expressions depending on the condition; in this case, the returned value would always be the same 'sb' object whereas the side-effect would be what we were originally aiming for: to append text to the StringBuilder.

Note that there is no need to return the 'sb', especially to assign it again. Therefore I suggest to have the ternary operator removed from here and have an if-else statement implemented instead.

Reply:
Or we can assign the message to a string variable (and continue to use the ternary operator, avoiding lots of ifs) and append it to the StringBuilder further down, before line 169.